[Challenge] Inverse battle challenge

KimJeongSun opened the pull request:
<!-- Make sure the title includes categorization (i.e. [Bug], [QoL], [Localization]) -->
<!-- Make sure that this PR is not overlapping with someone else's work -->
<!-- Please try to keep the PR self-contained (and small) -->

## What are the changes?
<!-- Summarize what are the changes from a user perspective on the application -->
implement an Inverse Battle Challenge.

## Why am I doing these changes?
<!-- Explain why you decided to introduce these changes -->
<!-- Does it come from an issue or another PR? Please link it -->
<!-- Explain why you believe this can enhance user experience -->
Many users in the Discord feature-vote channel have expressed a desire for this feature. (including myself!)

## What did change?
<!-- Explicitly state what are the changes introduced by the PR -->
<!-- You can make use of a comparison between what was the state before and after your PR changes -->
- Added the Inverse Battle Challenge.
- Refactored the `type.ts` file.
- add `isInverseBattleChallenge()` condition in move.ts to make `Thunder wave` not immuned against `Ground` type in Inverse Battle
- add `isInverseBattleChallenge()` condition in move.ts to make `Conversion 2` works in Inverse Battle
- add test code for inverse battle
- update Inverse Battle Challenge Achievement and Icon.
- change freeze dry code to support inverse battle challenge ( but marked partial requested by @Xavion3 )
- [excel inverse battle checklist link](https://docs.google.com/spreadsheets/d/1f52vkOhKsyi6yS-O1VhAndrSJufj3cLgNWpz3b55Zys/edit?usp=sharing)


### Screenshots/Videos
<!-- If your change is changing anything on the user experience, please provide visual proofs of it -->
<!-- Please take screenshots/videos before and after your changes, to show what is brought by this PR -->

- inverse battle example

https://github.com/user-attachments/assets/4d54aa6a-f58f-4152-a676-dc3feddc1ee8

- inverse battle challenge + Bug Type challenge

<img width="1685" alt="Screenshot 2024-08-17 at 6 19 04 PM" src="https://github.com/user-attachments/assets/a9b4fc4b-d0cd-4e10-b49d-9d5bb294b38f">

- ending video for ^ ( bug type challenge achievement is not activated )


https://github.com/user-attachments/assets/59e1cade-553a-4b20-b071-b163f6549d12

- after ending achivement ( there is no bug type challenge achivement )

<img width="1689" alt="Screenshot 2024-08-17 at 4 39 14 PM" src="https://github.com/user-attachments/assets/67f52ab2-11c1-4823-b497-3d46c1071250">



- block inverse battle challlenge + other challenges example ( edit starting wave to 220 floor for debugging )
  - mono gen + inverse battle challenge -> inverse battle challenge achievement only

https://github.com/user-attachments/assets/7d9d01f2-b731-4559-868d-b6981cc5a24e

  - mono type + fresh start + inverse battle challenge -> inverse battle challenge achievement only

https://github.com/user-attachments/assets/f1955bcc-cbd3-4b92-833f-3fbb6ed173a4



## How to test the changes?
<!-- How can a reviewer test your changes once they check out on your branch? -->
<!-- Did you just make use of the `src/overrides.ts` file? -->
<!-- Did you introduce any automated tests? -->
<!-- Do the reviewer need to do something special in order to test your change? -->

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [ ] Have I considered writing automated tests for the issue?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

flx-sta commented:
If you refactor code please also make sure to eliminate any outstanding `integer` and replace them with `number` (the real type, not some fake name)

KimJeongSun commented:
updated! but there are army of left `integer` on everywhere...

flx-sta commented:
> updated! but there are army of left `integer` on everywhere...

Yeah. But it's just an alias for `number` so we don't have to force it

flx-sta commented:
I think the name of the challenge should also be a bit more specific.
`inverse battle challenge` sounds a bit like I'd be fighting on the enemy side (top).
I would have expected something like `Inverse types challenge`

KimJeongSun commented:
Hmm, well, I didn't come up with this name, so I'm not sure.
https://bulbapedia.bulbagarden.net/wiki/Inverse_Battle

flx-sta commented:
> Hmm, well, I didn't come up with this name, so I'm not sure. https://bulbapedia.bulbagarden.net/wiki/Inverse_Battle

never mind then

Adri1 commented:
Possible to inclue other locales placeholders? 🙏

KimJeongSun commented:
@Adri1 It's still a draft PR, but thank you for taking an interest and leaving a comment.

Hmm, originally I was planning to finish developing the features needed for the first release and add them before marking it as ready to review, but if you prefer, we can proceed now.

First, I need to add some content regarding the English translations. ( After consulting with the balance team, we decided that for the first release, even if the existing challenges are combined with the inverse battle challenge, the existing challenge achievements will not be unlocked. We plan to specify this in the challenge description. )
After updating, I will inform the translation team and request their feedback and translations.

CodeTappert commented on src/locales/de/achv.ts:
```diff
@@ -272,6 +272,10 @@ export const PGMachv: AchievementTranslationEntries = {
     description: "Schließe die 'Neuanfang' Herausforderung ab"
+  },
+  "INVERSE_BATTLE": {
+    name: "Mirror rorriM",
+    description: "Complete the Inverse Battle challenge.\n.egnellahc elttaB esrevnI eht etelpmoC"
```
```suggestion
    name: "Spieglein, Spieglein an der Wand",
    description: "Schließe die 'Umkehrkampf' Herausforderung ab"
```

CodeTappert commented on src/locales/de/challenges.ts:
```diff
@@ -28,5 +28,11 @@ export const challenges: TranslationEntries = {
     "value.0": "Aus",
     "value.1": "An",
+  },
+  "inverseBattle": {
+    "name": "Inverse Battle",
+    "desc": "Type matchups are reversed and no type is immune to any other type.\nDisables other challenges' achievements.",
+    "value.0": "Off",
+    "value.1": "On",
```
```suggestion
  "inverseBattle": {
    "name": "Umkehrkampf",
    "desc": "Die Typen-Effektivität wird umgekehrt, und kein Typ ist gegen einen anderen Typ immun. Deaktiviert die Erfolge anderer Herausforderungen.",
    "value.0": "Aus",
    "value.1": "An",
```

CodeTappert replied:
This could be to long. Please check!

CodeTappert replied:
Also @flx-sta please check my translation

KimJeongSun replied:
<img width="1689" alt="image" src="https://github.com/user-attachments/assets/b7993f2b-71ed-48d1-87bc-24d763e93d27">

do i need to change font size?

CodeTappert replied:
Either that or i can make it shorter by using bullet points

KimJeongSun replied:
I think the font looks unnecessarily large in other languages as well. Let me adjust this first.

CodeTappert replied:
![image](https://github.com/user-attachments/assets/3072a509-5fb3-49bd-92a6-fd75fda7ac7e)

CodeTappert replied:
i will make a suggestion for the bullet points. Its good enough imo

CodeTappert replied:
> I think the font looks unnecessarily large in other languages as well. Let me adjust this first.

Ok tell me if you need the bullet point thing

KimJeongSun replied:
<img width="1689" alt="Screenshot 2024-08-17 at 6 09 43 PM" src="https://github.com/user-attachments/assets/bb201bf0-f7e5-4eeb-bc70-5ed77f4abb9b">
<img width="1689" alt="Screenshot 2024-08-17 at 6 11 05 PM" src="https://github.com/user-attachments/assets/c592b650-963f-4a7a-b076-4ecab689f321">

i think 2nd one looks simple, but 1st one looks more unified with other desc

KimJeongSun replied:
which way do you like?

CodeTappert replied:
> which way do you like?

do the first one. But remove the - after "Typen"

KimJeongSun replied:
or we can add `\n` in 1st option too

KimJeongSun replied:
```
    "desc": "Die Typen wird umgekehrt, und kein Typ ist gegen einen anderen Typ immun.\nDeaktiviert die Erfolge anderer Herausforderungen.",
```
is it correct?

CodeTappert replied:
i will create a suggestion

KimJeongSun replied:
<img width="1701" alt="image" src="https://github.com/user-attachments/assets/dd2ade02-c103-4bc6-9da3-7efc8ace8b5e">

KimJeongSun replied:
<img width="1702" alt="image" src="https://github.com/user-attachments/assets/372d8625-bf0f-4e00-a436-4b9c8eb1858d">

final version! how about it?

CodeTappert replied:
I hate that "Typen" and "Effektivität" are not on the same line... But thats just the fault of the german language for beeing to long :D 

We can keep it like this. Looks fine!

KimJeongSun replied:
It's unfortunate, but this seems to be the best we can do before adjusting the size of the explanatory UI window. 🤔 
Thank you for translating it quickly! 👍

KimJeongSun replied:
<img width="1691" alt="image" src="https://github.com/user-attachments/assets/0d4a7d5d-30e7-4992-a4d5-6efd76e7673e">

if devs allow me to change UI size, then this will be final version.

CodeTappert replied:
well in this case you can readd the "-" between "Typen" and "Effektivität"

KimJeongSun replied:
in the very last moment, i will double check to you.

KimJeongSun replied:
<img width="1689" alt="Screenshot 2024-08-17 at 7 12 21 PM" src="https://github.com/user-attachments/assets/8ae96f4d-b207-4c93-a4dd-544b7149b00d">

how about this last last final version?

CodeTappert commented on src/locales/de/challenges.ts:
```diff
@@ -28,5 +28,11 @@ export const challenges: TranslationEntries = {
+  },
+  "inverseBattle": {
+    "name": "Umkehrkampf",
+    "desc": "Die Typen-Effektivität wird umgekehrt, und kein Typ ist gegen einen anderen Typ immun. Deaktiviert die Erfolge anderer Herausforderungen.",
```
```suggestion
    "desc": "Die Typen Effektivität wird umgekehrt, und kein Typ ist gegen einen anderen Typ immun.\nDeaktiviert die Erfolge anderer Herausforderungen.",
```

Adri1 commented on src/locales/fr/challenges.ts:
```diff
@@ -28,5 +28,11 @@ export const challenges: TranslationEntries = {
     "value.1": "Oui",
+  },
+  "inverseBattle": {
+    "name": "Inverse Battle",
+    "desc": "Type matchups are reversed and no type is immune to any other type.\nDisables other challenges' achievements.",
+    "value.0": "Off",
+    "value.1": "On",
```
```suggestion
    "name": "Combat Inversé",
    "desc": "La table des types est inversée et disparition des immunités.\nDésactive les succès des autres challenges.",
    "value.0": "Non",
    "value.1": "Oui",
```

Adri1 commented on src/locales/fr/achv.ts:
```diff
@@ -271,6 +271,10 @@ export const PGMachv: AchievementTranslationEntries = {
     description: "Terminer un challenge « Nouveau départ »."
+  },
+  "INVERSE_BATTLE": {
+    name: "Mirror rorriM",
+    description: "Complete the Inverse Battle challenge.\n.egnellahc elttaB esrevnI eht etelpmoC"
```
```suggestion
    name: "La teté à verlan",
    description: "Terminer un challenge en Combat Inversé.\n..ésrevnI tabmoC ne egnellahc nu renimreT"
```

KimJeongSun replied:
<img width="1689" alt="Screenshot 2024-08-17 at 7 15 22 PM" src="https://github.com/user-attachments/assets/28141136-bf0b-4f73-87da-7c0ab567c30b">

how about it?

KimJeongSun replied:
plus this
<img width="1689" alt="Screenshot 2024-08-17 at 7 16 37 PM" src="https://github.com/user-attachments/assets/d1f61fe5-5472-44f0-9f26-c3a57dbe97f2">

AsdarDevelops commented on src/locales/es/challenges.ts:
```diff
@@ -27,5 +27,11 @@ export const challenges: TranslationEntries = {
     "value.1": "On",
+  },
+  "inverseBattle": {
+    "name": "Inverse Battle",
+    "desc": "Type matchups are reversed and no type is immune to any other type.\nDisables other challenges' achievements.",
+    "value.0": "Off",
+    "value.1": "On",
```
```suggestion
    "name": "Combate Inverso",
    "desc": "La efectividad de los tipos es invertida. No hay inmunidades entre tipos.\nEste reto deshabilita logros de otros retos.",
    "value.0": "Desactivado",
    "value.1": "Activado",
```

KimJeongSun replied:
can you suggest translation for achv.ts also?

KimJeongSun replied:
can you suggest Spanish translation for this too?
[src/locales/es/achv.ts](https://github.com/pagefaultgames/pokerogue/pull/3525/files#diff-fc5a0390b73e3f2ebedbc154e29f833130c6a2a96be30dae5a0002dcc9b409f2)

Adri1 commented on src/locales/fr/challenges.ts:
```diff
@@ -28,5 +28,11 @@ export const challenges: TranslationEntries = {
+  },
+  "inverseBattle": {
+    "name": "Combat Inversé",
+    "desc": "La table des types est inversée et disparition des immunités.\nDésactive les succès des autres challenges.",
```
```suggestion
    "desc": "Les affinités de la table des types sont inversées et plus aucun type n’a d’immunité.\nDésactive les succès des autres challenges.",
```

AsdarDevelops reviewed (commented):
Added Spanish

Adri1 reviewed (commented):
Tini French correction (thanks for the screenshot!)

Adri1 commented on src/locales/fr/achv.ts:
```diff
@@ -271,6 +271,10 @@ export const PGMachv: AchievementTranslationEntries = {
+  },
+  "INVERSE_BATTLE": {
+    name: "La teté à verlan",
+    description: "Terminer un challenge en Combat Inversé.\n..ésrevnI tabmoC ne egnellahc nu renimreT"
```
```suggestion
    description: "Terminer un challenge en Combat Inversé.\n.ésrevnI tabmoC ne egnellahc nu renimreT"
```

AsdarDevelops commented on src/locales/es/achv.ts:
```diff
@@ -271,6 +271,10 @@ export const PGMachv: AchievementTranslationEntries = {
     description: "Complete the Fresh Start challenge."
+  },
+  "INVERSE_BATTLE": {
+    name: "Mirror rorriM",
+    description: "Complete the Inverse Battle challenge.\n.egnellahc elttaB esrevnI eht etelpmoC"
```
```suggestion
    name: "Espejo ojepsE",
    description: "Completa el reto de Combate Inverso.\n.osrevnI etabmoC ed oter le atelpmoC"
```

AsdarDevelops reviewed (commented):
Missed the achv.ts!

CodeTappert commented:
How did you scale down the text?
I would love todo the same here:
![image](https://github.com/user-attachments/assets/75efb5b6-10d1-41d3-884b-eb2a59fc96d5)

Adri1 commented on src/locales/fr/achv.ts:
```diff
@@ -271,6 +271,10 @@ export const PGMachv: AchievementTranslationEntries = {
     description: "Terminer un challenge « Nouveau départ »."
+  },
+  "INVERSE_BATTLE": {
+    name: "La teté à verlan",
+    description: "Terminer un challenge en Combat Inversé.\n.ésrevnI tabmoC ne egnellahc nu renimreT"
```
```suggestion
    name: "La teuté à verlan",
    description: "Terminer un challenge en Combat Inversé.\nMineter un lenjcha en Ba-con Versin."
```

Adri1 reviewed (commented):
Better French adaptation

KimJeongSun commented:
> How did you scale down the text? I would love todo the same here: ![image](https://private-user-images.githubusercontent.com/38758606/358843018-75efb5b6-10d1-41d3-884b-eb2a59fc96d5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM5NjU1NjksIm5iZiI6MTcyMzk2NTI2OSwicGF0aCI6Ii8zODc1ODYwNi8zNTg4NDMwMTgtNzVlZmI1YjYtMTBkMS00MWQzLTg4NGItZWIyYTU5ZmM5NmQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE4VDA3MTQyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExNzFkOTFmMGQwYWEwNzcwNjM2YTkyODFkNzFmOTA4ODg3NmUzNGU5MDJiZWNmMmVkMTE4ZjczMTM5OWM5ZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1fi6Jvu_45p6jpulp9iGY4RspvoFa6c75yOsSPEVGXM)

<img width="1202" alt="Screenshot 2024-08-18 at 4 29 54 PM" src="https://github.com/user-attachments/assets/c58e86af-a019-470a-b7cf-839b258211f4">


@CodeTappert how about it? ( De font -> 80, middle align for achievement title label )

mercurius-00 commented on src/locales/zh_TW/achv.ts:
```diff
@@ -271,6 +271,10 @@ export const PGMachv: AchievementTranslationEntries = {
     description: "Complete the Fresh Start challenge."
+  },
+  "INVERSE_BATTLE": {
+    name: "Mirror rorriM",
+    description: "Complete the Inverse Battle challenge.\n.egnellahc elttaB esrevnI eht etelpmoC"
```
```suggestion
  "INVERSE_BATTLE": {
    name: "鏡子子鏡",
    description: "完成逆轉之戰挑戰\n戰挑戰之轉逆成完"
```

mercurius-00 commented on src/locales/zh_CN/challenges.ts:
```diff
@@ -28,4 +28,10 @@ export const challenges: TranslationEntries = {
     "value.1": "开启",
   },
+  "inverseBattle": {
+    "name": "Inverse Battle",
+    "desc": "Type matchups are reversed and no type is immune to any other type.\nDisables other challenges' achievements.",
+    "value.0": "Off",
+    "value.1": "On",
```
```suggestion
    "name": "逆转之战",
    "desc": "属性相克关系被反转，且没有任何属性对其他属性免疫。\n禁用其他挑战的成就。",
    "value.0": "关闭",
    "value.1": "开启",
```

mercurius-00 commented on src/locales/zh_TW/challenges.ts:
```diff
@@ -27,5 +27,11 @@ export const challenges: TranslationEntries = {
     "value.1": "On",
+  },
+  "inverseBattle": {
+    "name": "Inverse Battle",
+    "desc": "Type matchups are reversed and no type is immune to any other type.\nDisables other challenges' achievements.",
+    "value.0": "Off",
+    "value.1": "On",
```
```suggestion
    "name": "逆轉之戰",
    "desc": "屬性相克關系被反轉，且沒有任何屬性對其他屬性免疫。\n禁用其他挑戰的成就。",
    "value.0": "關閉",
    "value.1": "開啓",
```

mercurius-00 commented on src/locales/zh_CN/achv.ts:
```diff
@@ -271,6 +271,10 @@ export const PGMachv: AchievementTranslationEntries = {
     name: "初次尝试！",
     description: "完成初次尝试挑战"
+  },
+  "INVERSE_BATTLE": {
+    name: "Mirror rorriM",
+    description: "Complete the Inverse Battle challenge.\n.egnellahc elttaB esrevnI eht etelpmoC"
```
```suggestion
  "INVERSE_BATTLE": {
    name: "镜子子镜",
    description: "完成逆转之战挑战\n战挑战之转逆成完"
```

mercurius-00 reviewed (commented):
zh-cn and zh-tw

DayKev commented on src/data/type.ts:
```diff
@@ -257,299 +190,206 @@ export function getTypeDamageMultiplier(attackType: integer, defType: integer):
 }
 
 /**
  * Retrieve the types resisting a given type
```
```suggestion
 * Retrieve the types resisting a given type. Used by Conversion 2
```

DayKev commented on src/test/battle/inverse_battle.test.ts:
```diff
@@ -0,0 +1,245 @@
+  });
+
+  it("6. AI should use moves that are 2x effective - Fire Fang prefered over Ice Fang against Dragon Type", async () => {
+    game.override.seed("Fire Fang vs Ice Fang against Dragon Type");
```
```suggestion
```
Tests already use a fixed seed. Though if the AI sometimes doesn't pick Fire Fang based on seed then this test isn't very useful anyway.

KimJeongSun replied:
actually, it doesn't sometimes. can i remove this test 6?
because i am not sure there is way to compare AI score from test code?

DayKev replied:
Yeah the test should probably be removed since there's no good way to evaluate the AI's decisions (that I know of).

DayKev commented on src/test/utils/gameManager.ts:
```diff
@@ -137,6 +137,24 @@ export default class GameManager {
+      this.scene.pushPhase(new EncounterPhase(this.scene, false));
+      selectStarterPhase.initBattle(starters);
+    });
+
```
```suggestion

    if (overrides.OPP_HELD_ITEMS_OVERRIDE.length === 0) {
      this.game.removeEnemyHeldItems();
    }
    
```

KimJeongSun replied:
it looks suggested code looks not valid. do you think it is necessary?

DayKev replied:
Oh right, I forgot about imports. Make sure to add `import overrides from "#app/overrides.js";` at the top, and change this code to:
```ts
    if (overrides.OPP_HELD_ITEMS_OVERRIDE.length === 0) {
      this.removeEnemyHeldItems();
    }
```

flx-sta replied:
Without the `.js`

KimJeongSun commented on src/data/type.ts:
```diff
@@ -23,230 +23,163 @@ export enum Type {
+  let multiplier: number = 1;
   if (attackType === Type.UNKNOWN || defType === Type.UNKNOWN) {
-    return 1;
+    return multiplier = 1 as TypeDamageMultiplier;
```
it's my mis, i think it should be `multiplier as TypeDamageMultiplier;`
how about this?

DayKev replied:
Makes sense.

CodeTappert commented:
> > How did you scale down the text? I would love todo the same here: ![image](https://private-user-images.githubusercontent.com/38758606/358843018-75efb5b6-10d1-41d3-884b-eb2a59fc96d5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM5NjU1NjksIm5iZiI6MTcyMzk2NTI2OSwicGF0aCI6Ii8zODc1ODYwNi8zNTg4NDMwMTgtNzVlZmI1YjYtMTBkMS00MWQzLTg4NGItZWIyYTU5ZmM5NmQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE4VDA3MTQyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExNzFkOTFmMGQwYWEwNzcwNjM2YTkyODFkNzFmOTA4ODg3NmUzNGU5MDJiZWNmMmVkMTE4ZjczMTM5OWM5ZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1fi6Jvu_45p6jpulp9iGY4RspvoFa6c75yOsSPEVGXM)
> 
> <img alt="Screenshot 2024-08-18 at 4 29 54 PM" width="1202" src="https://private-user-images.githubusercontent.com/47556641/358890782-c58e86af-a019-470a-b7cf-839b258211f4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwNzAxODQsIm5iZiI6MTcyNDA2OTg4NCwicGF0aCI6Ii80NzU1NjY0MS8zNTg4OTA3ODItYzU4ZTg2YWYtYTAxOS00NzBhLWI3Y2YtODM5YjI1ODIxMWY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE5VDEyMTgwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiNmNkY2I1NWY5Y2UxMjNlMzI0OTZkOTQyOWQ4M2U4Y2Q0ZTIzMTU2MjI4NzI4OGIyN2IwZTJjMGU3N2MyNzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.EMDAvyWxXnIAghYevczkfLPQR9UXRm-HfOqvksKBat8">
> @CodeTappert how about it? ( De font -> 80, middle align for achievement title label )

Looks good! (Or i could just shorten the title tbh)

DayKev commented on src/data/type.ts:
```diff
@@ -257,299 +190,206 @@ export function getTypeDamageMultiplier(attackType: integer, defType: integer):
-  return 0;
+  // if isInverseBattle is true, reverse the multiplier
+  if (isInverseBattle) {
+    if (multiplier > 1) {
+      multiplier = 0.5;
+    } else if (multiplier < 1) {
+      multiplier = 2;
+    }
```
Ah right, forgot about immunities. That should work then.

josericardo-fo commented on src/locales/pt_BR/achv.ts:
```diff
@@ -273,6 +273,10 @@ export const PGMachv: AchievementTranslationEntries = {
     description: "Complete o desafio de novo começo."
   },
+  "INVERSE_BATTLE": {
+    name: "Mirror rorriM",
+    description: "Complete the Inverse Battle challenge.\n.egnellahc elttaB esrevnI eht etelpmoC"
```
```suggestion
    name: "Subi no ônibuS",
    description: "Complete o desafio da Batalha Invertida.\n.aditrevnI ahlataB ad oifased o etelpmoC"
```

josericardo-fo commented on src/locales/pt_BR/achv.ts:
```diff
@@ -548,4 +552,8 @@ export const PGFachv: AchievementTranslationEntries = {
     description: "Complete o desafio de novo começo."
   },
+  "INVERSE_BATTLE": {
+    name: "Mirror rorriM",
+    description: "Complete the Inverse Battle challenge.\n.egnellahc elttaB esrevnI eht etelpmoC"
```
```suggestion
    name: "Subi no ônibuS",
    description: "Complete o desafio da Batalha Invertida.\n.aditrevnI ahlataB ad oifased o etelpmoC"
```

josericardo-fo commented on src/locales/pt_BR/challenges.ts:
```diff
@@ -27,5 +27,11 @@ export const challenges: TranslationEntries = {
     "value.1": "Ligado",
+  },
+  "inverseBattle": {
+    "name": "Inverse Battle",
+    "desc": "Type matchups are reversed and no type is immune to any other type.\nDisables other challenges' achievements.",
+    "value.0": "Off",
+    "value.1": "On",
```
```suggestion
    "name": "Batalha Inversa",
    "desc": "Fraquezas e resistências de tipos são invertidas e nenhum tipo é imune a outro tipo.\nDesativa as conquistas de outros desafios.",
    "value.0": "Desligado",
    "value.1": "Ligado",
```

flx-sta commented on src/test/battle/inverse_battle.test.ts:
```diff
@@ -164,31 +163,13 @@ describe("Inverse Battle", () => {
-    await game.phaseInterceptor.to(MoveEndPhase);
-
-    expect(sandshrew.status?.effect).toBe(StatusEffect.PARALYSIS);
-  });
```
RIP?

KimJeongSun replied:
I understand that there will be a rework for that part in #3643 
So we decided to postpone updates for that part until after that PR is completed. (In my mind)

DayKev reviewed (changes requested):
Created a PR to fix up the tests, once you merge that I've got no complaints. https://github.com/KimJeongSun/pokerogue/pull/1

flx-sta commented on src/data/challenge.ts:
```diff
@@ -666,6 +680,41 @@ export class FreshStartChallenge extends Challenge {
+  /**
+   * @overrides
+   */
+  getDifficulty(): number {
```
NIT: you can use `override` as a keyword

```suggestion
  override getDifficulty(): number {
```

KimJeongSun replied:
i added override as a keyword and removed comments

flx-sta commented on src/data/challenge.ts:
```diff
@@ -666,6 +680,41 @@ export class FreshStartChallenge extends Challenge {
+    return false;
+  }
+
+  static loadChallenge(source: InverseBattleChallenge | any): InverseBattleChallenge {
+    const newChallenge = new InverseBattleChallenge();
+    newChallenge.value = source.value;
+    newChallenge.severity = source.severity;
+    return newChallenge;
+  }
+}
```
NIT: static method should be at the top of a class

KimJeongSun replied:
moved to top of the class method!

flx-sta commented on src/data/move.ts:
```diff
@@ -6154,6 +6157,25 @@ export class ResistLastMoveTypeAttr extends MoveEffectAttr {
+  getTypeResistances(gameMode: GameMode, type: number): Type[] {
+    const typeResistances: Type[] = [];
+
+    for (let i = 0; i < Type.STELLAR; i++) {
```
Q: 
Why `Type.STELLAR`? Is that the last one?

KimJeongSun replied:
changed to
`for (let i = 0; i < Object.keys(Type).length; i++) {`

flx-sta commented on src/locales/ca_ES/achv.ts:
```diff
@@ -271,6 +271,10 @@ export const PGMachv: AchievementTranslationEntries = {
     description: "Complete the Fresh Start challenge."
+  },
+  "INVERSE_BATTLE": {
+    name: "Mirror rorriM",
```
```suggestion
    name: "Mirror ɹoɹɹᴉW",
```

Idk if the font can render that :D

KimJeongSun replied:
do you want to change english translation?
( there is no ca_ES translation anymore? )

DayKev replied:
Yeah you can replace the English translations of these. Make sure it shows up properly in game, there might be font issues (as mentioned). If it breaks then just leave it as-is.

KimJeongSun replied:
i will try it!

KimJeongSun replied:
sadly, we should update font later

<img width="1714" alt="image" src="https://github.com/user-attachments/assets/00faacae-c777-4115-bbbe-c4cb65a45101">

flx-sta replied:
This looks good enough for me 😂

KimJeongSun replied:
art team will kill me 🔫 . plz stop

btw, i updated others too! can you review it again? 🙇

flx-sta commented on src/locales/ca_ES/achv.ts:
```diff
@@ -271,6 +271,10 @@ export const PGMachv: AchievementTranslationEntries = {
+  },
+  "INVERSE_BATTLE": {
+    name: "Mirror rorriM",
+    description: "Complete the Inverse Battle challenge.\n.egnellahc elttaB esrevnI eht etelpmoC"
```
```suggestion
    description: "Complete the Inverse Battle challenge.\n.ǝƃuǝllɐɥɔ ǝlʇʇɐq ǝsɹǝʌuI ǝɥʇ ǝʇǝldɯoƆ"
```
Idk if the font can render that :D

KimJeongSun replied:
do you want to change english translation?
( there is no ca_ES translation anymore? )

flx-sta commented on src/test/utils/gameManager.ts:
```diff
@@ -137,6 +138,28 @@ export default class GameManager {
 
   }
 
+  /**
+   * Runs the game to the summon phase.
+   * @param species - Optional array of species to summon.
+   * @returns A promise that resolves when the summon phase is reached.
+   */
+  async runToSummon(species?: Species[]) {
+    await this.runToTitle();
+
+    this.onNextPrompt("TitlePhase", Mode.TITLE, () => {
+      const starters = generateStarter(this.scene, species);
+      const selectStarterPhase = new SelectStarterPhase(this.scene);
+      this.scene.pushPhase(new EncounterPhase(this.scene, false));
+      selectStarterPhase.initBattle(starters);
+    });
+
+    if (overrides.OPP_HELD_ITEMS_OVERRIDE.length === 0) {
+      this.removeEnemyHeldItems();
+    }
+
+    await this.phaseInterceptor.run(EncounterPhase);
+  }
+
```
uuuh.. this was removed for good reason.
It was moved into `classicModeHelper.ts` and `dailyModeHelper.ts` instead

KimJeongSun replied:
Then, should I create a `challengeModeHelper.ts` for the inverse battle challenge test?
I am not sure how to proceed (whether I need to create complex functions to support each individual challenge).

DayKev replied:
I'd create a counterpart in `challengeModeHelper.ts` that starts the game in challenge mode instead of classic mode, and add some sort of `addChallenge()` function in that helper as well that has the various parameters of a challenge (your test uses `id`, `value`, and `severity` for example, though there might be more; and I assume some of them can have default values) that will then push that challenge to an internal `challenges` array. Then you can do:
```ts
game.challenge.addChallenge(id, ...);
game.challenge.startBattle(pokemon?);
```
and the `startBattle()` function of the challenge helper will load the game to the appropriate point, activate the challenges that have been specified, and then start the battle.

KimJeongSun replied:
I updated! can you review it again? @DayKev @flx-sta

flx-sta commented on src/test/utils/gameManager.ts:
```diff
@@ -174,8 +197,12 @@ export default class GameManager {
    */
-  async startBattle(species?: Species[]) {
-    await this.classicMode.runToSummon(species);
+  async startBattle(species?: Species[], isClassicMode = true) {
+    if (isClassicMode) {
+      await this.classicMode.runToSummon(species);
+    } else {
+      await this.runToSummon(species);
+    }
```
Please don't use a boolean for `isClassicMode`! instead move them into `classicModeHelper` and such.

You can mark the original method as deprecated if you don't want to change all the existing calls

KimJeongSun replied:
Then, should I define a runToSummon function (which corresponds to this.runToSummon in the previous code) inside the classicModeHelper.ts without designating the mode internally? Should it be defined as a separate class? Or just as a different name within the same class?
This function was used to circumvent the issue where the gameMode gets overridden by classic mode within the runToSummon function.

KimJeongSun replied:
I updated, can you review it again?

flx-sta reviewed (changes requested):
There were some changes made in testing utilities that shouldn't have been made!

flx-sta commented:
@KimJeongSun I was so kind and resolved those translation merge conflicts

CodeTappert commented:
> @KimJeongSun I was so kind and resolved those translation merge conflicts for you so you can focus on the review :)

But this is now missing the localized part in challenge.ts?

DayKev commented on src/test/utils/helpers/challengeModeHelper.ts:
```diff
@@ -0,0 +1,77 @@
+
+  /**
+   * Runs the Challenge game to the summon phase.
+   * @param species - Optional array of species to summon.
```
```suggestion
```
Some copy/paste leftovers to remove.

DayKev commented on src/test/utils/helpers/challengeModeHelper.ts:
```diff
@@ -0,0 +1,77 @@
+
+  /**
+   * Starts a battle in challenge mode.
+   * @param challenges - Array of challenges to set.
```
```suggestion
```
Same as other comment.

DayKev reviewed (commented):
Other than the minor comment copy-paste errors, lgtm.

Xavion3 commented on src/data/challenge.ts:
```diff
@@ -621,6 +635,13 @@ export class FreshStartChallenge extends Challenge {
     super(Challenges.FRESH_START, 1);
   }
 
+  static loadChallenge(source: InverseBattleChallenge | any): InverseBattleChallenge {
+    const newChallenge = new InverseBattleChallenge();
+    newChallenge.value = source.value;
+    newChallenge.severity = source.severity;
+    return newChallenge;
+  }
+
```
This is using the wrong type, and also why are you even adding a new loadChallenge function to fresh start? It already has one.

KimJeongSun replied:
I incorrectly pasted it while addressing the request to move the static function `loadChallenge` of flx to the top of the class. I apologize!!! I've fixed it.

Xavion3 commented on src/field/pokemon.ts:
```diff
@@ -1292,12 +1292,17 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
       }
-
-      return getTypeDamageMultiplier(moveType, defType);
+      const multiplier = new Utils.NumberHolder(1);
+      multiplier.value = getTypeDamageMultiplier(moveType, defType);
```
There is some redundancy here, do we want to just do `const multiplier = new Utils.NumberHolder(getTypeDamageMultiplier(moveType, defType));`?

KimJeongSun replied:
It's my first time using NumberHolder, so I didn't know. 
It's definitely better! I'll fix it.

Xavion3 commented on src/ui/achvs-ui-handler.ts:
Why are we changing the achievement UI in a PR about adding a new challenge?

KimJeongSun replied:
There was an issue where the title text of the added achievements and the text related to the Eviolite achievement did not display properly in German, so this part was inevitably modified. (I aligned the titles because I wanted to, and the German reviewer liked it. But if it becomes an issue, I can revert it.)

KimJeongSun replied:
opened new PR https://github.com/pagefaultgames/pokerogue/pull/3731

Xavion3 commented on src/ui/challenges-select-ui-handler.ts:
Why are we changing the challenge UI to add a new challenge, in ways unrelated to just adding a new row?

KimJeongSun replied:
The description of that achievement was a bit longer than others (to explain our deactivation decision of the achievement), causing the text to overflow out of the box. Therefore, I adjusted the aspect ratio to make it look more natural. At that time, the feedback from others was positive, but if I messed something up, I apologize.

Xavion3 replied:
It's just a general case, much like the achievements UI, that PRs should generally do one thing and a UI rework should really be seperate from adding a new challenge. A UI rework should have somewhat different standards than just adding a new challenge.

KimJeongSun replied:
Then should I just revert it to only changing the font size?

KimJeongSun replied:
opened new PR https://github.com/pagefaultgames/pokerogue/pull/3731

KimJeongSun commented:
close it, failed test

flx-sta commented:
> > @KimJeongSun I was so kind and resolved those translation merge conflicts for you so you can focus on the review :)
> 
> But this is now missing the localized part in challenge.ts?

Keen eyes. Thanks for letting me know.
I've restored those on my own behalf (as I also lost those).

https://github.com/pagefaultgames/pokerogue/pull/3525/commits/1c06faf4272191205032e8753d2e33723067bfc9

KimJeongSun commented:
@flx-sta @Xavion3  please check updated requested changes! 🙇

frutescens commented on src/test/battle/inverse_battle.test.ts:
```diff
@@ -0,0 +1,190 @@
+
+    expect(player.getTypes()[0]).toBe(Type.DRAGON);
+  }, TIMEOUT);
+});
```
Could you have some checks against abilities as well? Like to check that Pokemon with Fluffy/Dry Skin still take 2x damage against fire moves, etc. How does corrosion work in this mode? How does flying press (two types at once - flying and fighting) work?

DayKev replied:
Fluffy/etc don't interact with type matchups at all. Those abilities should have their own tests to see if they work (which if this PR does somehow break then those tests should properly break and there's no extra test needed here), that doesn't belong here.

Flying press should probably have a test though.

frutescens replied:
Yeah, that's fine by me. I meant that it would be good to make sure that those abilities shouldn't be affected by the type changes.

KimJeongSun replied:
I think I'll just add the test for Flying Press. (Honestly, I thought it would be fine to include it in the video.) 
I should be able to work on it this evening! I'll let you know once it's done, so please review and approve it if everything looks good!

torranx commented on src/data/challenge.ts:
```diff
@@ -327,6 +332,15 @@ export abstract class Challenge {
+  /**
+   * An apply function for TYPE_EFFECTIVENESS challenges. Derived classes should alter this.
+   * @param effectiveness {@link Utils.NumberHolder} The current effectiveness of the move.
+   * @returns {@link boolean} Whether this function did anything.
```
you can remove {@link boolean} here

KimJeongSun replied:
updated

torranx commented on src/data/challenge.ts:
```diff
@@ -327,6 +332,15 @@ export abstract class Challenge {
 
+  /**
+   * An apply function for TYPE_EFFECTIVENESS challenges. Derived classes should alter this.
+   * @param effectiveness {@link Utils.NumberHolder} The current effectiveness of the move.
```
nit on all occurrences of this: should be using `linkcode` instead of `link`

KimJeongSun replied:
I should be able to work on it this(+ remove link in return comment) evening! I'll let you know once it's done, so please review and approve it if everything looks good!

KimJeongSun replied:
updated!

frutescens commented:
I can update it myself through this PR but did you add Inverse Challenge to the Run History challenge parser?

KimJeongSun commented:
> I can update it myself through this PR but did you add Inverse Challenge to the Run History challenge parser?

I would be very grateful if you could open a PR and proceed with the merger into that PR. I'll go ahead and merge it with the latest beta for now.

frutescens commented:
Sent in my contribution. 
![image](https://github.com/user-attachments/assets/5a560ef8-2836-4ed8-a541-403dc54f8ac0)

KimJeongSun commented:
I would like to share a record with you about how I discovered an edge case with the move Freeze Dry in the inverse battle challenge while working on a test code for the move Flying Press by request from @frutescens 

First, I identified that the existing test code did not account for the additional type effectiveness of Flying Press, which includes the Flying type. The reason for this is that the following code only considers the basic type of the move, and it cannot check for type effectiveness changes due to additional effects:
```
expect(enemy.getAttackTypeEffectiveness(allMoves[Moves.FREEZE_DRY].type, player)).toBe(2);
```
To address this issue, I modified the test code from using `getAttackTypeEffectiveness` to the following:
```
expect(enemy.getMoveEffectiveness(player, allMoves[Moves.FREEZE_DRY])).toBe(2);
```
As a result, the move Flying Press now functions as expected and the test code works correctly.

Consequently, we discovered that in the context of an inverse battle matchup, the move Freeze Dry was actually functioning as an 8x multiplier against Water types! Here's the relevant code:
```
export class WaterSuperEffectTypeMultiplierAttr extends VariableMoveTypeMultiplierAttr {
  apply(user: Pokemon, target: Pokemon, move: Move, args: any[]): boolean {
    const multiplier = args[0] as Utils.NumberHolder;
    if (target.isOfType(Type.WATER)) {
      multiplier.value *= 4; // Increased twice because initial reduction against water
      return true;
    }

    return false;
  }
}
```
To correct this, I modified the implementation so that in an inverse matchup, the move does not get a 4x multiplier. The updated code is as follows:
```
export class WaterSuperEffectTypeMultiplierAttr extends VariableMoveTypeMultiplierAttr {
  apply(user: Pokemon, target: Pokemon, move: Move, args: any[]): boolean {
    const multiplier = args[0] as Utils.NumberHolder;
    if (target.isOfType(Type.WATER)) {
      const effectivenessAgainstWater = new Utils.NumberHolder(getTypeDamageMultiplier(move.getType(), Type.WATER));
      applyChallenges(user.scene.gameMode, ChallengeType.TYPE_EFFECTIVENESS, effectivenessAgainstWater);
      if (effectivenessAgainstWater.value < 1) {
        multiplier.value *= 4; // Increased twice because initial reduction against water
        return true;
      }
    }

    return false;
  }
}
```

flx-sta reviewed (dismissed):
This feels so wrong when playing with inverse types

Xavion3 commented on src/data/move.ts:
```diff
@@ -4178,8 +4189,12 @@ export class WaterSuperEffectTypeMultiplierAttr extends VariableMoveTypeMultipli
     if (target.isOfType(Type.WATER)) {
-      multiplier.value *= 4; // Increased twice because initial reduction against water
-      return true;
+      const effectivenessAgainstWater = new Utils.NumberHolder(getTypeDamageMultiplier(move.getType(), Type.WATER));
+      applyChallenges(user.scene.gameMode, ChallengeType.TYPE_EFFECTIVENESS, effectivenessAgainstWater);
+      if (effectivenessAgainstWater.value < 1) {
+        multiplier.value *= 4; // Increased twice because initial reduction against water
+        return true;
+      }
```
While more correct, this code doesn't quite work so would want a (P) or the like on the move. Freeze Dry should always deal super effective damage to water regardless of type matchup, and this code doesn't do that right now. Multiplying by 2/effectiveness will work though unless we ever add a way to make water immune to a type.

KimJeongSun replied:
I updated it

KimJeongSun commented on src/data/move.ts:
```diff
@@ -238,6 +241,14 @@ export default class Move implements Localizable {
     return this;
   }
 
+  /**
+   * get Move type
+   * @returns {@linkcode Type} the type of the move
+   */
+  getType(): Type {
+    return this._type;
+  }
+
```
Initially, I was planning to jut use Type.ICE, but since that attribute seemed to give the effect of being super effective against Water type, I prepared it to check for super effectiveness against general types. Of course, currently, that attribute is only being used for the Freeze-Dry move.

KimJeongSun replied:
I might have misunderstood the question or missed something, so I'm commenting again.
Is there another way to access the `this._type` (private) in the move class? I wasn't aware of it, so I created this function.

Tempo-anon commented:
I really love the idea of an inverse battle challenge modifier and think it would be great. However, inverse types are not very well balanced (normal is very OP) so would it make sense for it to interact or disable some mono type challenges? Beating inverse grass is very different from beating mono grass for example.

KimJeongSun commented:
> I really love the idea of an inverse battle challenge modifier and think it would be great. However, inverse types are not very well balanced (normal is very OP) so would it make sense for it to interact or disable some mono type challenges? Beating inverse grass is very different from beating mono grass for example.

So, I disabled it. (The balance team wanted it that way too!) :prayja:

Xavion3 commented:
Where is the code that is preventing the other achievements being earned during inverse battles?

frutescens reviewed (approved):
This was so difficult omg
![image](https://github.com/user-attachments/assets/7b7f2467-3d95-4589-9a8c-8b1926d58209)

frutescens reviewed (changes requested):
Not ready for merge. Please see second half of vidoe. 

https://github.com/user-attachments/assets/a714e771-3e8e-445d-8b1a-b3a62dd7beac

KimJeongSun commented:
> Not ready for merge. Please see second half of vidoe.
> 
>  freecompress-Screen.Recording.2024-08-27.at.8.27.41.PM.mp4

I found out blocking achv code in game over phase looks missing when we deleting old phases.ts
so i add it and for my test, works again!

frutescens reviewed (dismissed):
Issue resolved.

https://github.com/user-attachments/assets/970ed9b0-e839-4439-859f-21508a8b9789

frutescens commented on src/phases/game-over-phase.ts:
```diff
@@ -116,7 +117,16 @@ export class GameOverPhase extends BattlePhase {
           if (this.victory && this.scene.gameMode.isChallenge) {
-            this.scene.gameMode.challenges.forEach(c => this.scene.validateAchvs(ChallengeAchv, c));
+            // if it is inverse battle challenge, does not validate achvs except inverse battle achvs
+            if (!this.scene.gameMode.challenges.some(c => c.id === Challenges.INVERSE_BATTLE)) {
```
this code won't work because challenges are judged to be active based on their value -> if it's not 0 (on) or 0 (off)

KimJeongSun replied:
it is fixed!
